test/e2e/ingress.go

```diff
@@ -135,6 +143,32 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
 			By("should reject HTTP traffic")
 			ExpectNoError(pollURL(fmt.Sprintf("http://%v/", ip), "", lbPollTimeout, httpClient, true))
 
+			By("should have correct firewall rule for ingress")
+			fw, err := getFwForIngress(cloudConfig)
+			Expect(err).NotTo(HaveOccurred())
+			// Retrieve nodeTags from the node instance as the target tags for the ingress firewall rule
+			nodes := framework.GetReadySchedulableNodesOrDie(jig.client)
+			Expect(len(nodes.Items) > 0).Should(BeTrue())
+			nodeTags, err := getInstanceTags(cloudConfig, nodes.Items[0].Name)
+			Expect(err).NotTo(HaveOccurred())
+			// Getting ingress backend nodePorts. Current GCE ingress controller allows
+			// traffic to the default backend nodePort by default, so retrieve that as well.
+			nodePorts := []string{}
+			defaultSvc, err := jig.client.Core().Services(api.NamespaceSystem).Get(defaultBackendName)
+			Expect(err).NotTo(HaveOccurred())
+			nodePorts = append(nodePorts, strconv.Itoa(int(defaultSvc.Spec.Ports[0].NodePort)))
+			// Test service need to be consistent with the content in below file:
+			// test/e2e/testing-manifests/ingress/static-ip.svc.yaml
+			testSvc, err := jig.client.Core().Services(ns).Get("echoheaders-https")
+			Expect(err).NotTo(HaveOccurred())
+			nodePorts = append(nodePorts, strconv.Itoa(int(testSvc.Spec.Ports[0].NodePort)))
+			expFw, err := constructFwForIngress(fw.Name, nodePorts, nodeTags.Items)
+			Expect(err).NotTo(HaveOccurred())
+			// Passed the last argument as true to verify the backend ports are
+			// included but not exactly matched, given there may be other existing
+			// ingress resources and backends we are not aware of.
+			Expect(verifyFirewallRule(fw, expFw, cloudConfig.Network, true)).NotTo(HaveOccurred())
```

Comment:
So the test might become something like
```go
expectedTags := jig.getNodeTags()
expectedPorts := jig.getIngressNodePorts(ingName, ns)
gceController.verifyFirewall(ingName, expectedTags, expectedPorts)
```

which improves readibility by quite a bit, since I can just assume that the functions are doing what their names suggest, and understand the high level functionality we're testing by reading the `It()` block